pytorch_lightning/callbacks/progress.py

```diff
@@ -0,0 +1,371 @@
+    Example::
+
+        class LitProgressBar(ProgressBarBase):
+
+            def __init__(self):
+                super().__init__()  # don't forget this :)
+                self.enabled = True
+
+            def disable(self):
+                self.enabled = False
+
+            def on_batch_end(self, trainer, pl_module):
+                super().on_batch_end(trainer, pl_module)  # don't forget this :)
+                percent = (self.train_batch_idx / self.total_train_batches) * 100
+                sys.stdout.flush()
+                sys.stdout.write(f'{percent:.01f} percent complete \r')
+
+        bar = LitProgressBar()
+        trainer = Trainer(callbacks=[bar])
+
+    """
+    def __init__(self):
+
+        self._trainer = None
+        self._train_batch_idx = 0
+        self._val_batch_idx = 0
+        self._test_batch_idx = 0
+
+    @property
+    def trainer(self):
+        return self._trainer
+
+    @property
+    def train_batch_idx(self) -> int:
+        """
+        The current batch index being processed during training.
+        Use this to update your progress bar.
+        """
+        return self._train_batch_idx
+
+    @property
+    def val_batch_idx(self) -> int:
+        """
+        The current batch index being processed during validation.
+        Use this to update your progress bar.
+        """
+        return self._val_batch_idx
+
+    @property
+    def test_batch_idx(self) -> int:
+        """
+        The current batch index being processed during testing.
+        Use this to update your progress bar.
+        """
+        return self._test_batch_idx
+
+    @property
+    def total_train_batches(self) -> Optional[int]:
+        """
+        The total number of training batches during training, which may change from epoch to epoch.
+        Use this to set the total number of iterations in the progress bar. Can return ``inf`` if the
+        training dataloader is of infinite size.
+        """
+        if self.trainer.fast_dev_run:
+            total_train_batches = 1
+        else:
+            total_train_batches = self.trainer.num_training_batches
+        return total_train_batches
+
+    @property
+    def total_val_batches(self) -> Optional[int]:
+        """
+        The total number of training batches during validation, which may change from epoch to epoch.
+        Use this to set the total number of iterations in the progress bar. Can return ``inf`` if the
+        validation dataloader is of infinite size.
+        """
+        trainer = self.trainer
+        total_val_batches = 0
+        if trainer.fast_dev_run:
+            total_val_batches = len(trainer.val_dataloaders)
+        elif not self.trainer.disable_validation:
+            is_val_epoch = (trainer.current_epoch + 1) % trainer.check_val_every_n_epoch == 0
+            total_val_batches = trainer.num_val_batches if is_val_epoch else 0
+        return total_val_batches
+
+    @property
+    def total_test_batches(self) -> Optional[int]:
```

Comment:
i think that the `Oprional` stand for `Union[None, int]` not `Union[float, int]` so just int...

Reply:
Oh yes, because earlier I returned None when it was infinite number of batches. Good catch! will change

Reply:
so is the func bellow for converting inf -> None used?

Reply:
yes for tqdm later. but here just int instead of Optional[int] is correct as you pointed out.